src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilerCache.cs

```diff
@@ -44,157 +53,74 @@ internal CompilerCache(IEnumerable<RazorFileInfoCollection> viewCollections, IFi
+            var cachedResult = _cache.Get<Type>(normalizedPath);
+            if (cachedResult != null)
             {
-                return OnCacheMiss(relativeFileInfo, normalizedPath, compile, out result);
+                return CompilationResult.Successful(cachedResult);
             }
-            else
-            {
-                var fileInfo = relativeFileInfo.FileInfo;
-                if (cacheEntry.Length != fileInfo.Length)
-                {
-                    // Recompile if the file lengths differ
-                    return OnCacheMiss(relativeFileInfo, normalizedPath, compile, out result);
-                }
-
-                if (AssociatedViewStartsChanged(cacheEntry, compile))
-                {
-                    // Recompile if the view starts have changed since the entry was created.
-                    return OnCacheMiss(relativeFileInfo, normalizedPath, compile, out result);
-                }
 
-                if (cacheEntry.LastModified == fileInfo.LastModified)
-                {
-                    result = CompilationResult.Successful(cacheEntry.CompiledType);
-                    return cacheEntry;
-                }
+            var compilationResult = compile();
+            _cache.Set(normalizedPath,
+                       context => OnCacheMiss(context, compilationResult.CompiledType, relativePath));
 
-                // Timestamp doesn't match but it might be because of deployment, compare the hash.
-                if (cacheEntry.IsPreCompiled &&
-                    string.Equals(cacheEntry.Hash, RazorFileHash.GetHash(fileInfo), StringComparison.Ordinal))
-                {
-                    // Cache hit, but we need to update the entry.
-                    // Assigning to LastModified is an atomic operation and will result in a safe race if it is
-                    // being concurrently read and written or updated concurrently.
-                    cacheEntry.LastModified = fileInfo.LastModified;
-                    result = CompilationResult.Successful(cacheEntry.CompiledType);
-
-                    return cacheEntry;
-                }
-
-                // it's not a match, recompile
-                return OnCacheMiss(relativeFileInfo, normalizedPath, compile, out result);
-            }
+            return compilationResult;
         }
 
-        private CompilerCacheEntry OnCacheMiss(RelativeFileInfo file,
-                                               string normalizedPath,
-                                               Func<RelativeFileInfo, CompilationResult> compile,
-                                               out CompilationResult result)
+        private Type OnCacheMiss(ICacheSetContext context,
+                                 Type compiledType,
+                                 string relativePath)
         {
-            result = compile(file);
+            context.AddExpirationTrigger(_fileSystem.Watch(relativePath));
```

Comment:
Had a discussion with @Praburaj, @rynowak @Tratcher about this. We decided on starting with supporting the disk based checks for paths without wildcards to start with. I'll have the PR sent for file system next. This should alleviate issues for Mvc. We'll track replicating the behavior for paths with wildcards separately since it's a bit more involved.